components/unstoppable_domains/unstoppable_domains_navigation_throttle.cc

```diff
@@ -0,0 +1,93 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/unstoppable_domains/unstoppable_domains_navigation_throttle.h"
+
+#include <utility>
+
+#include "base/bind.h"
+#include "base/threading/thread_task_runner_handle.h"
+#include "brave/components/unstoppable_domains/unstoppable_domains_interstitial_controller_client.h"
+#include "brave/components/unstoppable_domains/unstoppable_domains_opt_in_page.h"
+#include "brave/components/unstoppable_domains/utils.h"
+#include "components/prefs/pref_service.h"
+#include "components/security_interstitials/content/security_interstitial_tab_helper.h"
+#include "components/user_prefs/user_prefs.h"
+#include "content/public/browser/browser_context.h"
+#include "content/public/browser/navigation_handle.h"
+#include "content/public/browser/web_contents.h"
+#include "net/base/net_errors.h"
+
+namespace unstoppable_domains {
+
+// static
+std::unique_ptr<UnstoppableDomainsNavigationThrottle>
+UnstoppableDomainsNavigationThrottle::MaybeCreateThrottleFor(
+    content::NavigationHandle* navigation_handle,
+    PrefService* local_state,
+    const std::string& locale) {
+  if (!IsUnstoppableDomainsEnabled())
+    return nullptr;
+
+  return std::make_unique<UnstoppableDomainsNavigationThrottle>(
+      navigation_handle, local_state, locale);
+}
+
+UnstoppableDomainsNavigationThrottle::UnstoppableDomainsNavigationThrottle(
+    content::NavigationHandle* navigation_handle,
+    PrefService* local_state,
+    const std::string& locale)
+    : content::NavigationThrottle(navigation_handle),
+      local_state_(local_state),
+      locale_(locale) {
+  content::BrowserContext* context =
+      navigation_handle->GetWebContents()->GetBrowserContext();
+  user_prefs_ = user_prefs::UserPrefs::Get(context);
+}
+
+UnstoppableDomainsNavigationThrottle::~UnstoppableDomainsNavigationThrottle() =
+    default;
+
+content::NavigationThrottle::ThrottleCheckResult
+UnstoppableDomainsNavigationThrottle::WillStartRequest() {
+  GURL url = navigation_handle()->GetURL();
+  if (!IsUnstoppableDomainsTLD(url) || !IsResolveMethodAsk(local_state_)) {
+    return content::NavigationThrottle::PROCEED;
+  }
+
+  base::ThreadTaskRunnerHandle::Get()->PostTask(
```

Comment:
AFAIK new official way to post tasks is `base::ThreadPool::PostTask`

Reply:
```

Reply:
```
[54437:70403:0212/102205.603336:FATAL:sequence_checker.h(120)] Check failed: checker.CalledOnValidSequence().
0   libbase.dylib                       0x00000001159a0d19 base::debug::CollectStackTrace(void**, unsigned long) + 9
1   libbase.dylib                       0x000000011587f293 base::debug::StackTrace::StackTrace() + 19
2   libbase.dylib                       0x00000001158a545f logging::LogMessage::~LogMessage() + 175
3   libbase.dylib                       0x00000001158a62fe logging::LogMessage::~LogMessage() + 14
4   libprefs.dylib                      0x00000001210f2436 PrefService::SetUserPrefValue(std::__Cr::basic_string<char, std::__Cr::char_traits<char>, std::__Cr::allocator<char> > const&, base::Value) + 86
5   libprefs.dylib                      0x00000001210f264f PrefService::SetBoolean(std::__Cr::basic_string<char, std::__Cr::char_traits<char>, std::__Cr::allocator<char> > const&, bool) + 47
6   libchrome_dll.dylib                 0x000000010ca8ba98 safe_browsing::UpdatePrefsBeforeSecurityInterstitial(PrefService*) + 168
7   libchrome_dll.dylib                 0x000000010cb63ab8 security_interstitials::SecurityInterstitialPage::SecurityInterstitialPage(content::WebContents*, GURL const&, std::__Cr::unique_ptr<security_interstitials::SecurityInterstitialControllerClient, std::__Cr::default_delete<security_interstitials::SecurityInterstitialControllerClient> >) + 104
8   libchrome_dll.dylib                 0x000000010b54bf80 unstoppable_domains::UnstoppableDomainsOptInPage::UnstoppableDomainsOptInPage(content::WebContents*, GURL const&, std::__Cr::unique_ptr<security_interstitials::SecurityInterstitialControllerClient, std::__Cr::default_delete<security_interstitials::SecurityInterstitialControllerClient> >) + 32
9   libchrome_dll.dylib                 0x000000010b54bb74 unstoppable_domains::UnstoppableDomainsNavigationThrottle::ShowInterstitial() + 164
```
So I guess we need to post a sequenced task here, I've changed to `base::SequencedTaskRunnerHandle::Get()->PostTask` here, sounds good?

Reply:
https://github.com/brave/brave-core/pull/7702/commits/1273796bfd5d774cb2a0f9e25d5c0c6846de4cf1